enterprise/app/tap/tap.tsx

```diff
@@ -120,12 +132,49 @@ export default class TapComponent extends React.Component<Props> {
     this.setState({
       loading: false,
       targetHistory: response.invocationTargets,
+      ...(this.isV2 && this.groupByCommit(response.invocationTargets)),
       stats: this.state.stats,
       maxInvocations: maxInvocations,
       maxDuration: maxDuration,
     });
   }
 
+  groupByCommit(targetHistories: target.ITargetHistory[]): Pick<State, "commits" | "commitToTargetIdToStatus"> {
+    const commitToMaxTargetEndTimestamp = new Map<string, number>();
+    const commitToTargetIdToStatus = new Map<string, Map<string, target.ITargetStatus>>();
+
+    for (const history of targetHistories) {
+      for (const targetStatus of history.targetStatus) {
+        const timestamp = endTimestampUnixNanos(targetStatus.timing);
+        const commitMaxTimestamp = commitToMaxTargetEndTimestamp.get(targetStatus.commitSha) || 0;
+        if (timestamp > commitMaxTimestamp) {
+          commitToMaxTargetEndTimestamp.set(targetStatus.commitSha, timestamp);
+        }
+
+        let targetIdToStatus = commitToTargetIdToStatus.get(targetStatus.commitSha);
+        if (!targetIdToStatus) {
+          targetIdToStatus = new Map<string, target.ITargetStatus>();
+          commitToTargetIdToStatus.set(targetStatus.commitSha, targetIdToStatus);
+        }
+
+        // For a given commit, the representative target status that
+        // we show in the UI is the one that ended the latest.
+        // TODO(bduffany): Keep track of per-target count by commit, in
+        // case the same target was executed multiple times for a given
+        // commit. Otherwise the count stat looks incorrect.
+        const existing = targetIdToStatus.get(history.target.id);
+        if (!existing || timestamp > endTimestampUnixNanos(existing.timing)) {
+          targetIdToStatus.set(history.target.id, targetStatus);
+        }
+      }
+    }
```

Comment:
@luluz66 FYI I'm re-ordering commits by latest `target.end_time_usec`, even though the server orders commits by `max(invocation.created_at_usec)`, since the server doesn't return the ordered commits list. LMK what you think -- I don't have a strong preference between the client or the server controlling the order of the rendered commits

Reply:
If we end up doing pagination server side, we probably want the ordering to match

Reply:
True, if the client just "concatenates" target histories and then recomputes these indexes from scratch each time, then the commits could wind up getting swapped around as new pages are fetched. It's probably better if the server returns an explicit list of commits

Reply:
I believe start_time_usec and end_time_usec of target might be wrong. They can be before invocation created_at _usec if the tests are cached. How do you feel about I pass the invocation's created_at_usec in the Target? When there are multiple invocations per commit, you also need the created_at_usec to show the last one right?

Reply:
> pass the invocation's created_at_usec in the Target

Discussed offline -- this is the option we decided to go with. Once it's implemented on the backend I'll update the ordering in this PR